### PR TITLE
Condensed down mavlink version detection

### PR DESCRIPTION
**Pull request by stephendade**
Simplified mavlink version options into a single commandline option

### CODE REVIEW

File: MAVProxy/mavproxy.py [hunk 1]
```diff
@@ -908,10 +908,7 @@ def run_script(scriptfile):
         action='append',
         default=[],
         help='Load the specified module. Can be used multiple times, or with a comma separated list')
-    parser.add_option("--mav09", action='store_true', default=False, help="Use MAVLink protocol 0.9")
-    parser.add_option("--mav10", action='store_true', default=False, help="Use MAVLink protocol 1.0")
-    parser.add_option("--mav20", action='store_true', default=True, help="Use MAVLink protocol 2.0")
-    parser.add_option("--auto-protocol", action='store_true', default=False, help="Auto detect MAVLink protocol version")
+    parser.add_option("--mavversion", default="", help="Force MAVLink Version (1.0, 2.0). Otherwise autodetect version")
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Should really use `1.0`` as the default here.

Should use choices([...]) to limit the options.

Also, it could force 0.9, so the comment's not quite right.

File: MAVProxy/mavproxy.py [hunk 2]
```diff
@@ -930,9 +927,9 @@ def run_script(scriptfile):
     if os.path.exists("/usr/sbin/ModemManager"):
         print("WARNING: You should uninstall ModemManager as it conflicts with APM and Pixhawk")
 
-    if opts.mav09:
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: I think we should try to preserve the old arguments for a time.  There are going to be people out there with scripts using the old arguments.

I suggest a subroutine `set_mav_version` or similar, taking opts as an argument.

Passing mav10 and mavversion at the same time should be fatal.  Passing mav09/mav10 should issue a warning.

**Comment by stephendade** [comment]
Fixed - I've added the old options back, with errors if they conflict with --mavversion. There's also a warning that the old options are depreciated if they are used.

File: MAVProxy/mavproxy.py [hunk 3]
```diff
@@ -523,7 +523,7 @@ def process_master(m):
         sys.stdout.flush()
         return
 
-    if m.first_byte and opts.auto_protocol:
+    if m.first_byte and (opts.mavversion == "" or opts.auto_protocol):
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Should use `is None`, not `== ""`; if no argument is given to MAVProxy that's the value it will have here.

File: MAVProxy/mavproxy.py [hunk 4]
```diff
@@ -861,6 +861,34 @@ def run_script(scriptfile):
             mpstate.console.writeln("-> %s" % line)
         process_stdin(line)
     f.close()
+    
+def set_mav_version(mav10, mav20, autoProtocol, mavversion):
+    '''Set the Mavlink version based on commandline options'''
+    if(mav10 == True or mav20 == True or autoProtocol == True):
+        print("Warning: Using depreciated --mav10, --mav20 or --auto-protocol options. Use --mavversion instead")
+    
+    #sanity check the options
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Unless you want different error messages here, this can be simplified.  Given only one of the options is set, you can increment a counter, and if that counter is >1 in the end you can throw the error.
e.g.
```
if maversion is not None:
  sources += 1
if mav10:
  sources += 1
.
.
```

File: MAVProxy/mavproxy.py [hunk 5]
```diff
@@ -908,10 +936,10 @@ def run_script(scriptfile):
         action='append',
         default=[],
         help='Load the specified module. Can be used multiple times, or with a comma separated list')
-    parser.add_option("--mav09", action='store_true', default=False, help="Use MAVLink protocol 0.9")
     parser.add_option("--mav10", action='store_true', default=False, help="Use MAVLink protocol 1.0")
-    parser.add_option("--mav20", action='store_true', default=True, help="Use MAVLink protocol 2.0")
+    parser.add_option("--mav20", action='store_true', default=False, help="Use MAVLink protocol 2.0")
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Sorry, I gave you a bum steer here.  Shouldn't *change* the default from `1.0` as part of this patch.  It should probably be changed, because it does cause issues defaulting to 2.0, but it shouldn't be part of this patch.

File: MAVProxy/mavproxy.py [hunk 6]
```diff
@@ -930,10 +958,9 @@ def run_script(scriptfile):
     if os.path.exists("/usr/sbin/ModemManager"):
         print("WARNING: You should uninstall ModemManager as it conflicts with APM and Pixhawk")
 
-    if opts.mav09:
-        os.environ['MAVLINK09'] = '1'
-    if opts.mav20 and not opts.mav10:
-        os.environ['MAVLINK20'] = '1'
+    #set the Mavlink version, if required
+    set_mav_version(opts.mav10, opts.mav20, opts.auto_protocol, opts.mavversion)
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Still think should just pass `opts` through here, but not that fussed.

File: MAVProxy/mavproxy.py [hunk 7]
```diff
@@ -548,7 +548,7 @@ def process_mavlink(slave):
     except socket.error:
         return
     try:
-        if slave.first_byte and opts.auto_protocol:
+        if slave.first_byte and (opts.mavversion == "" or opts.auto_protocol):
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: I think by this point we should be able to check whether the version has been set in the environment.  `opts` is only in scope because Python sucks.

Yeah, you're only doing what the code currently does in terms of scope, but if the thing is going to be thoroughly tsted it should be Made Right, I think.

**Comment by peterbarker** [comment]
Oh, please also look out for adding trailing whitespace; you might find an editor setting to highlight trailing whitespace.

File: MAVProxy/mavproxy.py (hunk 5 above)
Comment (stephendade) [inline MAVProxy/mavproxy.py]: Default is auto-detect via mavversion == None. Should (in theory) pick up the correct mavlink version without interaction from the user.

**Comment by stephendade** [comment]
I've tidied up the error handling and whitespace. Should be good now

**Comment by tridge** [comment]
hi @peterbarker, you ok with this now?

File: MAVProxy/mavproxy.py [hunk 8]
```diff
@@ -523,7 +526,7 @@ def process_master(m):
         sys.stdout.flush()
         return
 
-    if m.first_byte and opts.auto_protocol:
+    if m.first_byte and mavversion == False:
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Should be `None`

File: MAVProxy/mavproxy.py [hunk 9]
```diff
@@ -548,7 +551,7 @@ def process_mavlink(slave):
     except socket.error:
         return
     try:
-        if slave.first_byte and opts.auto_protocol:
+        if slave.first_byte and mavversion == False:
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Should be `None`

File: MAVProxy/mavproxy.py [hunk 10]
```diff
@@ -861,6 +864,32 @@ def run_script(scriptfile):
...
+
+    #and set the specific mavlink version (False = autodetect)       
+    if mavversion == "1.0" or mav10 == True:
+        os.environ['MAVLINK09'] = '1'
+        mavversion = "1"
+    elif mavversion == "2.0" or mav20 == True:
+        os.environ['MAVLINK20'] = '1'
+        mavversion = "2"
+    else:
+        mavversion = False
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: Should be `None`

File: MAVProxy/mavproxy.py [hunk 11]
```diff
@@ -861,6 +864,32 @@ def run_script(scriptfile):
             mpstate.console.writeln("-> %s" % line)
         process_stdin(line)
     f.close()
+    
+def set_mav_version(mav10, mav20, autoProtocol, mavversion):
+    '''Set the Mavlink version based on commandline options'''
+    if(mav10 == True or mav20 == True or autoProtocol == True):
+        print("Warning: Using depreciated --mav10, --mav20 or --auto-protocol options. Use --mavversion instead")
```
Comment (peterbarker) [inline MAVProxy/mavproxy.py]: `s/depreciated/deprecated/`

**Comment by stephendade** [comment]
Made changes as per comments by @peterbarker 
Also added global statements that I'd missed.

**Comment by peterbarker** [comment]
Pushed in.

Commented out the deprecation warning for the time being; tridge likes his `--mav10` :-)
